### PR TITLE
New commits function with "path" argument

### PR DESCRIPTION
**Pull request by pcarbo**
As you may know, we have been using the git2r package extensively for development of [workflowr](https://github.com/jdblischak/workflowr).

@jdblischak opened an issue a little while ago (Issue #324), and this is an attempt to add functionality to filter commits by file name---specifically, it recapitulates the behavior of `git log` when provided with the `<paths>` argument to return only commits in which the specified paths are modified.

Rather than modify the existing `commits` function (which could have unintended consequences), we implemented a new function `touching_commits` (the name could be changed) with fewer options.

Note that we are happy to make additional improvements, such as roxygen2 docs, after getting your initial feedback.

Finally, we have noticed that in some cases `touching_commits` does not exactly reproduce the behaviour of `git log <paths>`; it fails to identify some of the modifying commits for reasons we don't understand. We suspect that this is a lower level issue that goes beyond our understand of `libgit2`, and hopefully you will be able to identify the issue more quickly than us. (We can provide you with reproducible examples of this bug.)

### CODE REVIEW

**Comment by stewid** [comment]
Thanks, I review it within a few days.

**Comment by pcarbo** [comment]
Hi @stewid, do you think you might have a chance next week to look at our pull request? Even if it is only initial feedback, that would be helpful. We appreciate your contributions.

**Comment by stewid** [comment]
Hi @pcarbo, I now have some initial feedback. I think this
functionality would be a great addition to `git2r`.

In this first round, I haven't focused on the C code functionality:

1) I suggest that we drop the `touching_commits` function and instead
add the `path` argument to the `commits` function (I noticed that you
did that in the initial commit). We can then have something like

```
commits <- function(old arguments, path = NULL)
{
        .
        .
        .

     if (!is.null(path))
         return(.Call(git2r_revwalk_list2, repo, path))

    .Call(git2r_revwalk_list, repo, topological, time, reverse, n)
}
```

2) Could you please add roxygen2 docs for the `path` argument and one
example in the examples section.

3) We don't want to install the libgit2 example, I therefore suggest
that we remove the `inst/code` folder.

4) I strive to have a consistent C code style with 4 spaces for
indentation and /* */ for comments.

5) In `git2r_revwalk_list2`, add a check of the `path` argument and
open the repository with `git2r_repository_open`.

```
    if (git2r_arg_check_string(path))
        git2r_error(__func__, NULL, "'path'", git2r_err_string_arg);

    repository = git2r_repository_open(repo);
    if (!repository)
        git2r_error(__func__, NULL, git2r_err_invalid_repository, NULL);
```

6) Question: Now it is not possible to enter multiple paths, for
example, `commits(path = c("README.md", "README.Rmd"))`, is it
something we should try to fix??

7) If you provide me with a reproducible example of when it doesn't
work, I can try to debug the C code.

Thanks for your contribution

**Comment by pcarbo** [comment]
@stewid Thanks for your feedback. I will start working on these changes.

**Comment by pcarbo** [comment]
@jdblischak Can you please review what I had so far, and if it looks good, we can circle back to @stewid? Main improvements I made:

+ Clarified in docs that `path` argument only allows one file, and reproduces output of `git log --no-follow`.

+ Added examples illustrating use of `path` argument.

+ Cleaned up R and C++ code; `R CMD check --as-cran` passes.

**Comment by stewid** [comment]
Thanks for the pull request. Sorry for the late reply. I hope to find time tomorrow for review.

**Comment by pcarbo** [comment]
@stewid I'm not sure it is quite ready yet. @jdblischak will let you know it is ready for you to review. Thanks.

**Comment by jdblischak** [comment]
@stewid It is ready for review again. We addressed the items from your initial feedback and also fixed the bugs in our original implementation. Thanks!

**Comment by stewid** [comment]
Thanks

**Comment by stewid** [comment]
Hi @pcarbo @jdblischak 

I'm working on the review and I've made some minor changes. I'm thinking that instead of me copying the comments here, maybe it's easier that I create a PR in `pcarbo/master` that you can merge, before we merge this PR?  What do you think?

**Comment by pcarbo** [comment]
@stewid I invited you to `pcarbo/git2r`. So you are welcome to either submit a PR, or push directly to master, whichever you think is best. Thanks!

**Comment by jdblischak** [comment]
The CI builds have passed:

https://travis-ci.org/ropensci/git2r/builds/541437585
https://ci.appveyor.com/project/karthik/git2r/builds/25043945

@stewid @pcarbo Is there anything remaining to add for this PR? More tests, documentation, etc.?

@stewid Do you want to write the `NEWS` entry?

**Comment by pcarbo** [comment]
@pcarbo I have nothing else to add at this point.

File: src/git2r_revwalk.c [hunk 1]
```diff
@@ -215,7 +420,7 @@ SEXP git2r_revwalk_contributions(
     if (LOGICAL(topological)[0])
         sort_mode |= GIT_SORT_TOPOLOGICAL;
     if (LOGICAL(time)[0])
-        sort_mode |= GIT_SORT_TIME;
+        sort_mode = GIT_SORT_TIME | (sort_mode & GIT_SORT_REVERSE);
```
Comment (stewid) [inline src/git2r_revwalk.c]: Does this fix some issue that you have identified? I cannot see that this code is used by the `commits` function?

**Comment by stewid** [comment]
- I've added a NEWS entry in a PR
- I've also added a comment/question in the `git2r_revwalk_contributions` function.

Besides that, I have no further things to add to this PR before merging.

Thanks

**Comment by pcarbo** [comment]
@stewid Thanks! I merged your PR. Please merge this one whenever you are ready.

File: src/git2r_revwalk.c (hunk 1 above)
Comment (jdblischak) [reply src/git2r_revwalk.c]: @pcarbo Was this change necessary to get the commits to sort properly? You made the change in commit https://github.com/pcarbo/git2r/commit/068dd6a92b1c519e5f1d7ea95f507f2a6dca3425

Comment (pcarbo) [reply src/git2r_revwalk.c]: @stewid @jdblischak I got it from [here](https://github.com/libgit2/libgit2/blob/master/examples/log.c). I don't quite understand how `sort_mode` works, and I have no reason to believe that it works any better than `sort_mode |= GIT_SORT_TIME`. How you have it now in `HEAD` is fine.

Comment (jdblischak) [reply src/git2r_revwalk.c]: I wrote some tests for the arguments `time`,  `topological`, and `reverse`. I found both of the lines above to behave equivalently. I'm preparing a PR with the new tests and returning the `sort_mode` line to the briefer version.

Comment (jdblischak) [reply src/git2r_revwalk.c]: https://github.com/pcarbo/git2r/pull/5

**Comment by stewid** [comment]
Thanks @pcarbo @jdblischak for adding this feature to git2r

**Comment by jdblischak** [comment]
@stewid What's your tentative timeline for the next release of git2r? It's not super urgent for me, but I'd like to be able to plan ahead for when this new feature will be available from the CRAN version.

**Comment by stewid** [comment]
@jdblischak I'll try to find time next week to submit a new version to CRAN.

**Comment by stewid** [comment]
git2r v0.26.0 on CRAN now

**Pull request merged**